rust-src/concordium_base/src/base.rs

```diff
@@ -0,0 +1,1239 @@
+}
+
+pub type MintDistribution<CPV> = <CPV as MintDistributionFamily>::Output;
+
+#[derive(Debug, Serialize, Clone, Copy)]
+pub struct MintRate {
+    pub mantissa: u32,
+    pub exponent: u8,
+}
+
+#[derive(
+    Default,
+    Debug,
+    Display,
+    Clone,
+    Copy,
+    SerdeSerialize,
+    SerdeDeserialize,
+    Serialize,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Into,
+    FromStr,
+)]
+#[serde(transparent)]
+/// A fraction of an amount with a precision of `1/100_000`.
+/// The [`FromStr`] instance will parse a decimal fraction with up to `5`
+/// decimals.
+pub struct AmountFraction {
+    pub(crate) parts_per_hundred_thousands: PartsPerHundredThousands,
+}
+
+impl AmountFraction {
+    /// Construct a new fraction given the integer number of parts per
+    /// `100_000`. Return [`None`] if the number of parts exceeds `100_000`.
+    pub fn new(parts: u32) -> Option<Self> {
+        let parts_per_hundred_thousands = PartsPerHundredThousands::new(parts)?;
+        Some(Self {
+            parts_per_hundred_thousands,
+        })
+    }
+
+    /// Construct a new fraction, but does not check that the resulting fraction
+    /// is valid.
+    pub fn new_unchecked(parts: u32) -> Self {
+        Self {
+            parts_per_hundred_thousands: PartsPerHundredThousands::new_unchecked(parts),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, SerdeSerialize, SerdeDeserialize, Serialize, FromStr)]
+#[serde(transparent)]
+#[repr(transparent)]
+/// A bound on the relative share of the total staked capital that a baker can
+/// have as its stake. This is required to be greater than 0.
+pub struct CapitalBound {
+    #[serde(deserialize_with = "crate::internal::deserialize_non_default::deserialize")]
+    pub bound: AmountFraction,
+}
+
+#[derive(SerdeSerialize, SerdeDeserialize, Serialize)]
+#[serde(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug, FromStr, Display, From, Into)]
+/// Sequential index of finalization.
+pub struct FinalizationIndex {
+    pub index: u64,
+}
+
+/// Add two parts, checking that the result is still less than 100_000.
+impl std::ops::Add for PartsPerHundredThousands {
+    type Output = Option<Self>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let parts = self.parts.checked_add(rhs.parts)?;
+        if parts <= 100_000 {
+            Some(PartsPerHundredThousands { parts })
+        } else {
+            None
+        }
+    }
+}
+
+/// Add two reward fractions checking that they sum up to no more than 1.
+impl std::ops::Add for AmountFraction {
+    type Output = Option<Self>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let parts_per_hundred_thousands =
+            (self.parts_per_hundred_thousands + rhs.parts_per_hundred_thousands)?;
+        Some(AmountFraction {
+            parts_per_hundred_thousands,
+        })
+    }
+}
+
+impl SerdeSerialize for PartsPerHundredThousands {
+    /// FIXME: This instance needs to be improved and tested.
```

Comment:
fixme

Reply:
Obsolete. There are tests of this now, so I removed it.